pkg/runtime/nodejs/nodejs.go

```diff
@@ -0,0 +1,240 @@
+// Copyright 2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package nodejs
+
+import (
+	"debug/elf"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"strconv"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/prometheus/procfs"
+
+	"github.com/parca-dev/parca-agent/pkg/runtime"
+)
+
+var nodejsIdentifyingSymbols = [][]byte{
+	[]byte("InterpreterEntryTrampoline"),
+}
+
+func IsV8(ef *elf.File) (bool, error) {
+	return runtime.HasSymbols(ef, nodejsIdentifyingSymbols)
+}
+
+func IsRuntime(proc procfs.Proc) (bool, error) {
+	exe, err := proc.Executable()
+	if err != nil {
+		return false, err
+	}
+
+	var isNodeJS bool
+	if isNodeJSBin(exe) {
+		var err error
+		ef, err := elf.Open(absolutePath(proc, exe))
+		if err != nil {
+			return false, fmt.Errorf("open elf file: %w", err)
+		}
+
+		isNodeJS, err = runtime.HasSymbols(ef, nodejsIdentifyingSymbols)
+		if err != nil {
+			return false, fmt.Errorf("failed to check for symbols: %w", err)
+		}
+	}
+
+	if isNodeJS {
+		return true, nil
+	}
+
+	maps, err := proc.ProcMaps()
+	if err != nil {
+		return false, fmt.Errorf("error reading process maps: %w", err)
+	}
+
+	var (
+		found bool
+		lib   string
+	)
+	for _, m := range maps {
+		if pathname := m.Pathname; pathname != "" {
+			if m.Perms.Execute {
+				if isNodeJSLib(pathname) {
+					found = true
+					lib = pathname
+					break
+				}
+			}
+		}
+	}
+	if !found {
+		// If we didn't find a library, we can't be sure that this is a nodejs process.
+		return false, nil
+	}
+
+	ef, err := elf.Open(absolutePath(proc, lib))
+	if err != nil {
+		return false, fmt.Errorf("open elf file: %w", err)
+	}
+
+	isNodeJS, err = runtime.HasSymbols(ef, nodejsIdentifyingSymbols)
+	if err != nil {
+		return false, fmt.Errorf("failed to check for symbols: %w", err)
+	}
+
+	return isNodeJS, nil
+}
+
+func RuntimeInfo(proc procfs.Proc) (*runtime.Runtime, error) {
+	isNodeJS, err := IsRuntime(proc)
+	if err != nil {
+		return nil, fmt.Errorf("is runtime: %w", err)
+	}
+	if !isNodeJS {
+		return nil, nil //nolint:nilnil
```

Comment:
nit: could this result in a misuse of the API? I think most would expect a nil error to carry a non-nil result. Perhaps a custom error / return value would be worth it? (I assume that this is what the lint above tries to catch)

Reply:
This is how we designed the providers in the first place. We could of course always change. The idea is some providers are opt-in and wouldn't contribute any labels to the label set of the given process.